azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/PartitionManager.java

```diff
@@ -95,339 +123,370 @@ void onPartitionCheckCompleteTestHook()
-        this.host.stopExecutor();
-
-        TRACE_LOGGER.info(LoggingUtils.withHost(this.host.getHostName(), "Partition manager exiting"));
-
-    	return null;
+    			else
+    			{
+    	    		TRACE_LOGGER.warn(LoggingUtils.withHost(this.host,
+    	                    "Exception while initializing stores, not starting partition manager"), notifyWith);
+    			}
+    		}
+    	}, this.host.getExecutorService())
+    	// Stage 3: schedule scan, which will find partitions and start pumps, if previous stages succeeded
+    	.thenRunAsync(() ->
+    	{
+			// Schedule the first scan right away.
+    		synchronized (this.scanFutureSynchronizer)
+    		{
+    			this.scanFuture = this.host.getExecutorService().schedule(() -> scan(), 0, TimeUnit.SECONDS);
+    		}
+	    	
+			onInitializeCompleteTestHook();
+    	}, this.host.getExecutorService());
     }
     
-    private void initializeStores() throws InterruptedException, ExecutionException, ExceptionWithAction, IllegalEntityException
+    private CompletableFuture<?> initializeStores()
     {
         ILeaseManager leaseManager = this.host.getLeaseManager();
-        
-        // Make sure the lease store exists
-        if (!leaseManager.leaseStoreExists().get())
-        {
-        	retryWrapper(() -> leaseManager.createLeaseStoreIfNotExists(), null, "Failure creating lease store for this Event Hub, retrying",
-        			"Out of retries creating lease store for this Event Hub", EventProcessorHostActionStrings.CREATING_LEASE_STORE, 5);
-        }
-        // else
-        //	lease store already exists, no work needed
-        
-        // Now make sure the leases exist
-        for (String id : getPartitionIds())
-        {
-        	retryWrapper(() -> leaseManager.createLeaseIfNotExists(id), id, "Failure creating lease for partition, retrying",
-        			"Out of retries creating lease for partition", EventProcessorHostActionStrings.CREATING_LEASE, 5);
-        }
-        
         ICheckpointManager checkpointManager = this.host.getCheckpointManager();
         
-        // Make sure the checkpoint store exists
-        if (!checkpointManager.checkpointStoreExists().get())
+        // Stage 0: does lease store exist?
+        CompletableFuture<?> initializeStoresFuture = leaseManager.leaseStoreExists()
+        // Stage 1: create lease store if it doesn't exist
+        .thenComposeAsync((exists) ->
         {
-        	retryWrapper(() -> checkpointManager.createCheckpointStoreIfNotExists(), null, "Failure creating checkpoint store for this Event Hub, retrying",
-        			"Out of retries creating checkpoint store for this Event Hub", EventProcessorHostActionStrings.CREATING_CHECKPOINT_STORE, 5);
-        }
-        // else
-        //	checkpoint store already exists, no work needed
+        	CompletableFuture<?> createLeaseStoreFuture = null;
+        	if (!exists)
+        	{
+        		createLeaseStoreFuture = retryWrapper(() -> leaseManager.createLeaseStoreIfNotExists(), null,
```

Comment:
>retryWrapper [](start = 35, length = 12)

why do we retry here?
Since, this is run as part of initialize - how about giving early feedback to the user? thats what .net client does.

Reply:
So that one transient Storage failure doesn't kill EPH startup? Each operation is retried five times, and if it runs out of retries, then startup fails immediately and it will bubble up to user from the register API.